packages/appium/docs/en/ecosystem/build-plugins.md

```diff
@@ -2,4 +2,436 @@
+metadata'):
+
+    ```json
+    {
+      ...,
+      "appium": {
+        "pluginName": "fake",
+        "mainClass": "FakePlugin"
+      },
+      ...
+    }
+    ```
+
+The required subfields are:
+
+* `pluginName`: this should be a short name for your plugin.
+* `mainClass`: this is a named export (in CommonJS style) from your `main` field. It must be a
+  class which extends Appium's `BasePlugin` (see below).
+
+### Extend Appium's `BasePlugin` class
+
+Ultimately, your plugin is much easier to write because most of the hard work of defining patterns
+for overriding commands is done for you. This is
+all encoded up as a class which Appium exports for you to use, called `BasePlugin`. It is exported
+from `appium/plugin`, so you can use one of these styles to import it and create your *own* class
+that extends it:
+
+```js
+import {BasePlugin} from 'appium/plugin';
+// or: const {BasePlugin} = require('appium/plugin');
+
+export class MyPlugin extends BasePlugin {
+}
+```
+
+### Make your plugin available
+
+That's basically it! With a Node.js package exporting a plugin class and with correct Appium
+extension metadata, you've got yourself an Appium plugin! Now it doesn't *do* anything, but you can
+load it up in Appium, activate it, etc...
+
+To make it available to users, you could publish it via NPM. When you do so, your plugin will be
+installable via the Appium CLI:
+
+```
+appium plugin install --source=npm <plugin-package-on-npm>
+```
+
+It's a good idea to test your plugin first, of course. One way to see how it works within Appium is
+to install it locally first:
+
+```
+appium plugin install --source=local /path/to/your/plugin
+```
+
+And of course, plugins must be "activated" during Appium server start, so make sure you direct your
+users to do so:
+
+```
+appium --use-plugins=plugin-name
+```
+
+### Developing your plugin
+
+How you develop your plugin is up to you. It is convenient, however, to run it from within Appium
+without having to do lots of publishing and installing. The most straightforward way to do this is
+to include the most recent version of Appium as a `devDependency` (although its being already
+included as a `peerDependency` is sufficient in newer versions of NPM), and then also your own plugin,
+like this:
+
+```json
+{
+    "devDependencies": {
+        ...,
+        "appium": "^2.0.0",
+        "your-plugin": "file:.",
+        ...
+    }
+}
+```
+
+Now, you can run Appium locally (`npm exec appium` or `npx appium`), and because your plugin is
+listed as a dependency alongside it, it will be automatically "installed" and available. You can
+design your e2e tests this way, or if you're writing them in Node.js, you can simply import
+Appium's start server methods to handle starting and stopping the Appium server in Node.
+
+Another way to do local development with an existing Appium server install is to simply install
+your plugin locally:
+
+```
+appium plugin install --source=local /path/to/your/plugin/dev/dir
```

Comment:
isn't this repeating the previous paragraph?

Reply:
yeah, will remove.